core/shared/src/main/scala/zio/prelude/fx/ZPure.scala

```diff
@@ -877,6 +878,23 @@ object ZPure extends ZPureLowPriorityImplicits with ZPureArities {
   def accessM[R]: AccessMPartiallyApplied[R] =
     new AccessMPartiallyApplied
 
+  def fromEffectCatching[E >: Null <: Throwable]: FromEffectCatchingPartiallyApplied[E] =
```

Comment:
I think that implementing `refineOrDie` on `ZPure` so that we could do `fromEffect(???).refineOrDie[NumberFormatException]` would be a more compositional way of addressing this. This would also be consistent with the way this use case would be handled with ZIO.

Reply:
I tried to implement it, but we're not in ZIO. ZPure doesn't have a "die" error channel, so the only option would be to literally `throw` and that doesn't seem to me like a good idea :shrug:

Reply:
Maybe it should have a "die" channel??? There is an issue that we're still on a particular platform so let's say we are interpreting a `ZPure` value with an error type of `String` and we evaluate a continuation that throws an `IllegalArgumentException`. Right now we have nowhere to put that so the only thing we can do no is throw it, which isn't the end of the world but throws away any information we have about other errors that occurred as well as ability of higher levels of the application to handle it other than by explicitly using `try` and `catch`.

Reply:
:thinking: That's an interesting idea!

I haven't given it much thought, but I think ZPure could do well without "die channel". ZIO does need it, but ZPure isn't _"ZIO with more type parameters"_, so it could (should?) be fine without it. Maybe the best ZPure can do is to pretend "unchecked" exceptions don't exist. And if one is raised, which should be in a rare, truly serious and exceptional case, it will blow right through ZPure. This approach could make the design and intended use of ZPure more focused, which could be a good thing.

But from me, this is still just a hunch; no strong feeling either way. So if your guts tell you to go that way, go ahead. We can try it see how it works and perhaps even get some feedback from users :+1:

Reply:
I would err on the side of no channel for now. It's true this can result in the loss of the "primary error" (the `E`). Whereas, ZIO itself would hold onto the `E` and the `Throwable` via `Cause`. But ZPure is more likely to be used for short, pure computations, where exhaustive error reporting has less value; which is why I don't think adding stack traces, etc., is absolutely essential for ZPure (unlike ZIO).